### PR TITLE
ENH: Add in sort keyword to DatetimeIndex.union

### LINKED ISSUE #24994: 0.24.0 DateTimeIndex: union() got an unexpected keyword argument 'sort'
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
index1 = pd.DataFrame(index=pd.to_datetime(['19900101'])).index
index2 = index1.copy()
index1.union(index2, sort=False)
TypeError: union() got an unexpected keyword argument 'sort'
```
#### Problem description
The 'sort' argument works with 'difference' and 'intersection' but fails on 'union' for DateTimeIndex.

### PR DESCRIPTION
**Pull request by reidy-p**
- [x] closes #24994, xref #24471
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

As discussed in #25063 for ``Index.union``, I have made the default ``sort=None`` which should sort except in the three cases outlined in the doc-string. When passing ``sort=False`` it should never sort.

There is a fastpath in the ``DatetimeIndex.union`` that at present automatically sorts the two ``DatetimeIndexes`` if they are both monotonic by simply putting the two indexes in the correct order and then removing any overlaps. I have tried to adjust this fastpath to control this sorting behaviour.

I have tried to adjust the tests to properly test the sorting behaviour but I think the set operations tests need more work in general.

@jorisvandenbossche and @TomAugspurger it would be great if you could have a look at this too.

### CODE REVIEW

File: pandas/core/indexes/datetimes.py [hunk 1]
```diff
@@ -469,15 +469,27 @@ def union(self, other):
...
+            Whether to sort the resulting Index.
+
+            * None : Sort the result, except when
+
+              1. `self` and `other` are equal.
+              2. `self` or `other` has length 0.
+              3. Some values in `self` or `other` cannot be compared.
+                 A RuntimeWarning is issued in this case.
+
+            * False : do not sort the result
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: add a versionaded tag

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche if you have a quick look.

File: doc/source/whatsnew/v0.25.0.rst [hunk 2]
```diff
@@ -22,6 +22,7 @@ Other Enhancements
 - Indexing of ``DataFrame`` and ``Series`` now accepts zerodim ``np.ndarray`` (:issue:`24919`)
 - :meth:`Timestamp.replace` now supports the ``fold`` argument to disambiguate DST transition times (:issue:`25017`)
 - :meth:`DataFrame.at_time` and :meth:`Series.at_time` now support :meth:`datetime.time` objects with timezones (:issue:`24043`)
+- :meth:`DatetimeIndex.union` now supports the ``sort`` argument. The behaviour of the sort parameter matches that of ``Index.union`` (:issue:`24994`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you add a meth reference here

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. very minor comment in whatsnew. ping on green.

**Comment by reidy-p** [comment]
@jreback it's green now. Thanks!

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@reidy-p thanks a lot!

**Comment by jorisvandenbossche** [comment]
@reidy-p also, feel free to update https://github.com/pandas-dev/pandas/issues/24471 with the current status (do now all index classes have a sort keyword for intersection?)

**Pull request merged**